docs/blog/2024-08-27-challenges-of-microservices.md

```diff
@@ -0,0 +1,152 @@
+5. Finally, the `notifications` service notifies the customer that the parcel is on its way.
+
+![Simple e-commerce workflow](./2024-08-27-simple-checkout.png)
+
+### Microservices are Distributed
+
+Recall that each service (in the workflow diagram above, each box) is its own deployable artifact. That means that the happy-path business process described above will involve five different software systems from start-to-finish.
+
+In the above workflow diagram, each arrow can be accurately interpreted in two ways:
+1. The logical flow of the business process.
+2. The physical flow of information between microservices, either through network RPC calls or through a message broker like Apache Kafka.
+
+Guess what! This means we have a distributed system by definition. As Splunk [writes in a blog post](https://www.splunk.com/en_us/blog/learn/distributed-systems.html):
+> A distributed system is simply any environment where multiple computers or devices are working on a variety of tasks and components, all spread across a network.
+
+You need to look no further than the [Fallacies of Distributed Computing](https://en.wikipedia.org/wiki/Fallacies_of_distributed_computing) (written by Sun Microsystems Fellow L. Peter Deutsch in 1994) to see that this means that microservices are no easy task.
+
+### Microservices are Leaderless
+
+As we've seen already, any microservice-based application is a distributed system. Some distributed systems have the concept of a _leader_, which is a special node in the system that has special responsibilities.
+
+:::info
+Apache Kafka is my favorite distributed system. In Apache Kafka, the _Controller_ is a special Kafka server that is responsible for deciding which partition replicas are hosted on (and led by) which brokers. If the broker who was in charge of a partition goes down, then the Controller chooses a new broker from the ISR to take its place.
+
+Therefore, the _Controller_ in Apache Kafka can be thought of as a _leader_.
+:::
+
+While systems like Apache Kafka have clear leaders (for example, the _Controller_ may re-assign partition leadership if the cluster becomes too imbalanced), in a microservice-based system there is no central leader to ensure that the chips fall correctly. 
+
+You can think of our e-commerce microservice flow as a line of dominoes falling. Once the process starts, no one entity is responsible for ensuring its completion. The business workflow moves from `orders` to `inventory` to `payments` and so on. If `payments` fails for some reason (perhaps a network partition makes the Stripe API unavailable), then it's quite possible that the `shipping` service never finds out about the workflow.
+
+However, in real life such outcomes are not acceptable. This means that every single player in the system must:
+
+1. Have built-in reliability mechanisms.
+2. Understand the preceding and subsequent steps of the business process to route traffic.
+
```

Comment:
What is the problem with adding these to each of these?  Why would having to code in #1 and #2 be bad?

Suggestion:  Implementing the above slows down development, more tightly couples one services to another, increases dependancies, and negates the advantages of a microservices archiecture.